text/0037-node-upgrades.md

```diff
@@ -0,0 +1,98 @@
+# Node Upgrades
+
+## Summary
+
+[summary]: #summary
+
+CEP PR: [casperlabs/ceps#0037](https://github.com/casperlabs/ceps/pull/0037)
+
+This CEP describes the how the `casper-node-launcher` and `casper-node` will interact to achieve network upgrades.
+
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+The `casper-node-launcher` is a relatively simple Rust binary which runs the `casper-node` binary as a child process.
+Normal operation will see the launcher running indefinitely, cycling upwards through versions of the node as and when
+they become active.
+
+The node has two relevant modes of operation: "validator mode", and "migrate-data mode".  The normal mode is validator.
+The migrate-data mode is used immediately after an upgrade to migrate any data stored on disk as required.
+
+The node also has three possible exit codes: 0, 101 or 102.  0 indicates success, 101 indicates an error (e.g. a panic)
+and 102 indicates the node should be downgraded to a previous version.  These exit codes are read and reacted to by the
+launcher.
+
+On startup, the launcher either tries to read its previously cached state from disk, or assumes a fresh start.
+Information cached includes which mode of operation the node was executed in, the protocol version, the path to the node
+binary and the path to the config folder.
+
+On a fresh start, the launcher searches for the lowest installed version of `casper-node` and starts running it in
+validator mode.
+
+If the launcher cannot find an appropriate version at any stage of upgrading or downgrading, it exits with an error.
+
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+### casper-node-launcher
+
+The following describes the control flow in the launcher:
+
+![control flow in the launcher](images/0037/casper-launcher-flow.png "control flow in the launcher")
+
+### casper-node
+
+On startup, the node needs to decide what action to take depending upon whether it is an initial run (without fast sync
+this will always be treated as running from genesis), a restart after an upgrade, or a restart after an unexpected
```

Comment:
With fast sync this will still be treated as running from genesis if a trusted hash is not provided...